server2/app.py

```diff
@@ -0,0 +1,333 @@
+    token = rest
+    if initial_path:
+        token[oauth_initial_path_key] = initial_path
+    if oauth_token:
+        token[oauth_token_key] = oauth_token
+    if oauth_state:
+        token[oauth_state_key] = oauth_state
+    if sub:
+        token['sub'] = sub
+    token['jti'] = str(uuid.uuid1())
+    token['iss'] = base_uri
+    token['aud'] = base_uri
+    token['exp'] = expires
+    token['iat'] = now
+    token['nbf'] = now
+    logger.debug('JWT params: %s', token)
+    return token
+
+
+class Cookie:
+    def __init__(self):
+        self.keys = {}
+        self.ages = {}
+        self.secures = {}
+        self.deletes = []
+
+    def remove_jwt(self, key):
+        self.deletes.append(key)
+
+    def add_jwt(self, key, jwt, max_age=300, secure=True):
+        self.keys[key] = jwt
+        self.ages[key] = max_age
+        self.secures[key] = secure
+
+    def populate_resp(self, resp):
+        for key, token in self.keys.items():
+            resp.set_cookie(
+                key=key,
+                max_age=self.ages[key],
+                secure=self.secures[key],
+                value=jwt.encode(
+                    token,
+                    # The first secret is the current signer.
+                    jwt_secrets[0],
+                    algorithm=algorithm
+                )
+            )
+        for key in self.deletes:
+            resp.set_cookie(
+                key=key,
+                max_age=0
+            )
+        return resp
+
+
+def get_jwt_token(cookies, key):
+    """
+    Retrieve the jwt token's as a key, value object
+    from the flask request cookies object.
+    If no token exists, returns an empty object.
+    """
+    if key in request.cookies:
+        # We take multiple secrets to allow for online secret rotation.
+        # The first secret is the current signer,
+        # and the others are potentially still in use, but will
+        # be rotated out.
+        for jwt_secret in jwt_secrets:
+            try:
+                # github.com/jpadilla/pyjwt/blob/master/tests/test_api_jwt.py
+                # This will validate:
+                #   iss, aud
+                # This will validate (if they are in the request):
+                #   exp, nbf, iat
+                return jwt.decode(
+                    request.cookies[key],
+                    jwt_secret,
+                    issuer=base_uri,
+                    audience=base_uri,
+                    algorithms=[algorithm]
+                )
+            except jwt.exceptions.InvalidTokenError as e:
+                # will catch exp, nbf, iat, iss, aud errors
+                print(e)
+                continue
+    return {}
+
+
+def with_auth(func):
+    """
+    Authentication decorator.
+
+    Run the given request handler, passing it the user.
+    If the user has no valid auth token, then
+    redirect the user/resource owner to the keymaster
+    without invoking the given request handler.
+    """
+    @wraps(func)
```

Comment:
This is more aesthetic, but this decorator name is meh. perhaps importing it as something more reasonable, like 'authcheck' or something